### PR TITLE
Taxonomy-aware filtering

### PR DESCRIPTION
**Pull request by laughedelic**
A significant amount of 16S sequences are lost with the current filter. As an example, this RNACentral entry
- http://rnacentral.org/rna/URS00005C321D/243230

has as header in the FASTA file

```
>URS00005C321D Deinococcus radiodurans R1 rrn
```

The sequences we are going to keep are those which
1. have rRNA in the `rna_type` field
2. their taxonomy assignment is 
   1. not one of the blacklisted ids we had, _and_
   2. a descendant of either Bacteria (tax id 2) or Archaea (tax id 2157)
3. are longer than 1300

### CODE REVIEW

**Comment by laughedelic** [comment]
This is done. I put the check that uses Bio4j in the end as the most expensive (so if any of the previous conditions is not met, it won't touch the DB).

@eparejatobes review, please.

**Comment by eparejatobes** [comment]
@laughedelic done, please re-review 83249f1, merge, release and launch! :hammer: :smiley:

File: src/main/scala/rna16s.scala [hunk 1]
```diff
@@ -36,27 +36,34 @@ case object bio4jTaxonomyBundle extends AnyBio4jDist {
...
-        case None => false
-        case Some(parent) =>
-          if (ancestors.contains( parent.id() )) true
-          else checkAncestors_rec(parent)
-      }
-
-    optional(graph.nCBITaxonIdIndex.getVertex(id))
-      .map(checkAncestors_rec)
-      .getOrElse(false)
+  implicit class IdOps(val id: String) extends AnyVal {
```
Comment (laughedelic) [inline src/main/scala/rna16s.scala:39]: I don't see much point in this dancing over adding one method to _String_... But whatevs..

File: src/main/scala/rna16s.scala [hunk 2]
```diff
@@ -91,29 +102,28 @@ case object rna16s extends AnyBlastDB {
...
+    ( ! uninformativeTaxIDs.contains(row.select(tax_id)) )  &&
+    /* - the corresponding sequence is not shorter than the threshold */
+    (fasta.getV(sequence).value.length >= minimum16SLength) &&
     /* - is a descendant of either Archaea or Bacteria */
-    bio4jTaxonomyBundle.checkAncestors(
-      row.select(tax_id), Set(
+    row.select(tax_id).isDescendantOfOneIn(
+      Set(
         "2157", // Archaea
         "2"     // Bacteria
```
Comment (laughedelic) [inline src/main/scala/rna16s.scala:122]: So length `1300` has to be a named constant, but these poor bacterias don't deserve it... 😝

**Comment by laughedelic** [comment]
OK. I'm going to merge it.

**Comment by laughedelic** [comment]
Done. 526 681 sequences after filtering. 

```
$ ls -FshR1 outputs/
outputs/:
total 857M
4.0K blastdb/
835M era7bio.db.rna16s.fasta
 22M id2taxa.tsv

outputs/blastdb:
total 274M
 69M era7bio.db.rna16s.fasta.nhr
6.1M era7bio.db.rna16s.fasta.nin
199M era7bio.db.rna16s.fasta.nsq
```

One thing, the log has these lines:

```
===> makeblastdb -in /media/ephemeral0/outputs/era7bio.db.rna16s.fasta -input_type fasta -dbtype nucl -title era7bio.db.rna16s
FASTA-Reader: Ignoring invalid residues at position(s): On line 8615784: 25
FASTA-Reader: Ignoring invalid residues at position(s): On line 8652011: 25
FASTA-Reader: Ignoring invalid residues at position(s): On line 8652077: 25
```

This is a bit strange, because before I didn't see these messages. Probably it's just some new sequences. I've checked the 8615784 line:

> GCCCCGGUAAACGGCGGCCGUAAC**X**AUAACGGUCCUAAGGUAGCGAAAUUCCUUGUCGGGUAAGUUCCGA

So it has **X** at position 25..

@eparejatobes WDYT?

**Pull request merged**